### PR TITLE
Add support of RPC protocol in `Storage` layer

### PR DESCRIPTION
**Pull request by Eagle941**
This PR replaces the Pathfinder database interface with a Starkware RPC client supporting v0.7.

Techical details:
- `Pathfinder` dependency has been removed
- The trait `Storage` has been updated with types coming from `starkware-api` crate
- Ignored tests are now runnable in CI/CD because they don't rely on the Pathfinder database in the disk
- The blockifier is called directly instead of relying on a Starkware node for transaction execution
- The RPC client has been implemented using the crate `jsonrpc`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,7 +42,7 @@ This tool makes use of `tracing` library for log purposes. For this reason set
 ## Example
 
 ```bash
-cargo run -- --db-path ../pathfinder/mainnet.sqlite --start-block 632917 --end-block 632917 --svg-out "histogram.svg"
+cargo run --release -- --rpc-url https://starknet-mainnet.public.blastapi.io/rpc/v0_7 --start-block 632917 --end-block 632917 --svg-out "histogram.svg"
```
Comment (iamrecursion) [inline README.md:45]: Is this a public endpoint that people will be able to hit if they just run this command?

File: cli/src/args.rs [hunk 2]
```diff
@@ -4,12 +4,15 @@
 use std::path::PathBuf;
 
 use clap::Parser;
+use url::Url;
 
+/// This is the struct of the command line arguments accepted by
+/// `starknet-replay`.
 #[derive(Clone, Parser, Debug)]
 pub struct Args {
     /// The path of the Pathfinder database file.
```
Comment (iamrecursion) [inline cli/src/args.rs:13]: This is not correct any more.

File: starknet-replay/Cargo.toml [hunk 3]
```diff
@@ -21,29 +21,38 @@ cairo-lang-compiler = "2.6.0"
 cairo-lang-sierra = "2.6.0"
 cairo-lang-utils = "2.6.0"
 cairo-lang-sierra-generator = "2.6.0"
-# Some changes are required to `eqlabs/pathfinder` and to
-# `starkware-libs/blockifier`. These changes are in the branch `extract_libfunc`
-# of Reilabs fork and need to be merged in main branch.
+# Some changes are required to `starkware-libs/blockifier`. These changes are
+# in the branch `extract_libfunc` of Reilabs fork and need to be merged
+# in main branch.
```
Comment (iamrecursion) [inline starknet-replay/Cargo.toml:26]:
```suggestion
# in the branch `extract_libfunc` of Reilabs' fork and need to be merged
# into the main branch.
```

File: starknet-replay/src/error/runner.rs [hunk 4]
```diff
@@ -51,24 +47,30 @@ pub enum Error {
         start_block: BlockNumber,
     },
 
-    /// `CastError` variant is triggered when casting from `u64` to `usize`.
+    /// `IntTooBig` variant is triggered when casting from `u64` to `usize`
+    /// returns an error.
     #[error(transparent)]
-    CastError(#[from] TryFromIntError),
+    IntTooBig(#[from] TryFromIntError),
```
Comment (iamrecursion) [inline starknet-replay/src/error/runner.rs:53]: Maybe something like `IntegerTooLarge` would read a bit more nicely.

File: starknet-replay/src/profiler/analysis.rs [hunk 5]
```diff
@@ -165,10 +169,10 @@ mod tests {
             .unwrap_or_else(|_| panic!("Unable to read file {sierra_program_json_file}"));
         let sierra_program_json: serde_json::Value = serde_json::from_str(&sierra_program_json)
             .unwrap_or_else(|_| panic!("Unable to parse {sierra_program_json_file} to json"));
-        let contract_class: SierraContractClass =
-            serde_json::from_value::<SierraContractClass>(sierra_program_json).unwrap_or_else(
-                |_| panic!("Unable to parse {sierra_program_json_file} to SierraContractClass"),
-            );
+        let contract_class: ContractClass =
+            serde_json::from_value::<ContractClass>(sierra_program_json).unwrap_or_else(|_| {
```
Comment (iamrecursion) [inline starknet-replay/src/profiler/analysis.rs:173]: Surely the annotation on the let binding should obviate the need for the turbofish here?

File: starknet-replay/src/runner/mod.rs [hunk 6]
```diff
@@ -147,22 +121,17 @@ pub fn get_visited_program_counters(
...
 #[must_use]
 pub fn process_transaction_traces(
-    transaction_simulations: Vec<TransactionSimulation>,
-    block_number: BlockNumber,
+    transaction_simulations: Vec<(TransactionExecutionInfo, VisitedPcs)>,
 ) -> VisitedPcs {
     let mut cumulative_visited_pcs = VisitedPcs::default();
     for simulation in transaction_simulations {
-        let Some(visited_pcs) = get_visited_program_counters(simulation.trace) else {
+        // TODO: this is a performance hit. Refactor to avoid cloning.
```
Comment (iamrecursion) [inline starknet-replay/src/runner/mod.rs:128]: This `TODO` doesn't mention an issue. Please file one and mention it here.

File: starknet-replay/src/runner/replay_class_hash.rs [hunk 7]
```diff
@@ -2,14 +2,17 @@
...
 use starknet_api::core::ClassHash as StarknetClassHash;
 
 use crate::runner::BlockNumber;
 
 /// [`ReplayClassHash`] combines [`StarknetClassHash`] with
 /// [`crate::block_number::BlockNumber`] in order to uniquely identify a
 /// Contract Class from the database.
-#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash, PartialOrd, Ord)]
+#[derive(
+    Debug, Default, Copy, Clone, Eq, PartialEq, Hash, PartialOrd, Ord, Deserialize, Serialize,
```
Comment (iamrecursion) [inline starknet-replay/src/runner/replay_class_hash.rs:14]: Please put derives in alphabetical order.

File: starknet-replay/src/runner/replay_state_reader.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
+//! This module implements [`blockifier::state::state_api::StateReader`] for use
+//! in Starknet transactions replay. The calls to read the blockchain state rely
```
Comment (iamrecursion) [inline starknet-replay/src/runner/replay_state_reader.rs:2]:
```suggestion
//! in Starknet transaction replay. The calls to read the blockchain state rely
```

File: starknet-replay/src/runner/replay_state_reader.rs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+use starknet::core::types::ContractClass as StarknetContractClass;
+use starknet_api::core::{ClassHash, CompiledClassHash, ContractAddress, Nonce};
+use starknet_api::hash::StarkFelt;
+use starknet_api::state::StorageKey;
+
+use crate::block_number::BlockNumber;
+use crate::runner::replay_class_hash::ReplayClassHash;
+use crate::storage::rpc::{contract_class, RpcStorage};
+
+pub struct ReplayStateReader<'a> {
```
Comment (iamrecursion) [inline starknet-replay/src/runner/replay_state_reader.rs:17]: Missing doc block for the struct.

File: starknet-replay/src/runner/replay_state_reader.rs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    /// The block number used to query the state.
+    block_number: BlockNumber,
+}
+impl ReplayStateReader<'_> {
+    /// Constructs a new [`ReplayStateReader`] object.
+    ///
+    /// # Arguments
+    ///
+    /// - `storage`: The reference object exposing RPC calls to query blockchain
```
Comment (iamrecursion) [inline starknet-replay/src/runner/replay_state_reader.rs:29]: You don't need to say "reference object". The type is part of the documentation, and it is clear that it is a reference.

File: starknet-replay/src/runner/report.rs [hunk 11]
```diff
@@ -20,15 +21,15 @@ use crate::error::RunnerError;
 /// Returns [`Err`] if there is any error writing to `filename`.
 pub fn write_to_file(
     filename: &PathBuf,
-    traces: &Vec<TransactionSimulation>,
+    traces: &Vec<(TransactionExecutionInfo, VisitedPcs)>,
```
Comment (iamrecursion) [inline starknet-replay/src/runner/report.rs:24]: Make this output type a struct.

File: starknet-replay/src/storage/rpc/class_info.rs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+use crate::storage::Storage;
+
+/// This internal function returns the
+/// [`blockifier::execution::contract_class::ClassInfo`] for both Sierra and
+/// CASM contracts.
+///
+/// # Arguments
+///
+/// - `storage`: the struct which implements the trait
+///   [`crate::storage::Storage`] to query the Starknet blockchain.
```
Comment (iamrecursion) [inline starknet-replay/src/storage/rpc/class_info.rs:25]: Saying the type is unnecessary as the generated doc has the function signature. Same below.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Good stuff, but some changes are needed for clarity.

File: README.md (hunk 1 above)
Comment (Eagle941) [reply README.md:45]: The endpoint is known public starknet rpc server and everybody can use it

File: cli/src/main.rs [hunk 13]
```diff
@@ -4,8 +4,14 @@
...
-#![warn(clippy::all, clippy::cargo, clippy::pedantic)]
-#![allow(clippy::multiple_crate_versions)] // Due to duplicate dependencies in pathfinder
+#![warn(
+    clippy::all,
+    clippy::cargo,
+    clippy::pedantic,
+    clippy::missing_docs_in_private_items,
+    clippy::unwrap_used
+)]
+#![allow(clippy::multiple_crate_versions)] // Different libraries depend on different versions of the same library.
```
Comment (iamrecursion) [inline cli/src/main.rs:14]: This isn't a helpful comment. _Which_ libraries? Can we fix it?

**Review by iamrecursion — DISMISSED** [review]
LGTM barring one small nit.

File: cli/src/main.rs (hunk 13 above)
Comment (Eagle941) [reply cli/src/main.rs:14]: It's these many libraries:
```
multiple versions for dependency `base64`: 0.13.1, 0.21.7, 0.22.1
multiple versions for dependency `bitflags`: 1.3.2, 2.6.0
multiple versions for dependency `convert_case`: 0.4.0, 0.6.0
multiple versions for dependency `darling`: 0.14.4, 0.20.9
multiple versions for dependency `darling_core`: 0.14.4, 0.20.9
multiple versions for dependency `darling_macro`: 0.14.4, 0.20.9
multiple versions for dependency `hashbrown`: 0.12.3, 0.13.2, 0.14.5
multiple versions for dependency `indexmap`: 1.9.3, 2.2.6
multiple versions for dependency `itertools`: 0.10.5, 0.11.0
multiple versions for dependency `num-traits`: 0.1.43, 0.2.19
multiple versions for dependency `parking_lot`: 0.11.2, 0.12.3
multiple versions for dependency `parking_lot_core`: 0.8.6, 0.9.10
multiple versions for dependency `redox_syscall`: 0.2.16, 0.5.2
multiple versions for dependency `serde_with`: 2.3.3, 3.8.3
multiple versions for dependency `serde_with_macros`: 2.3.3, 3.8.3
multiple versions for dependency `starknet-core`: 0.5.1, 0.10.0
multiple versions for dependency `starknet-crypto`: 0.5.2, 0.6.2
multiple versions for dependency `starknet-curve`: 0.3.0, 0.4.2
multiple versions for dependency `strsim`: 0.10.0, 0.11.1
multiple versions for dependency `syn`: 1.0.109, 2.0.70
multiple versions for dependency `toml_edit`: 0.21.1, 0.22.15
multiple versions for dependency `windows-sys`: 0.48.0, 0.52.0
multiple versions for dependency `windows-targets`: 0.48.5, 0.52.6
multiple versions for dependency `windows_aarch64_gnullvm`: 0.48.5, 0.52.6
multiple versions for dependency `windows_aarch64_msvc`: 0.48.5, 0.52.6
multiple versions for dependency `regex-syntax`: 0.6.29, 0.8.4
multiple versions for dependency `windows_i686_gnu`: 0.48.5, 0.52.6
multiple versions for dependency `windows_i686_msvc`: 0.48.5, 0.52.6
multiple versions for dependency `windows_x86_64_gnu`: 0.48.5, 0.52.6
multiple versions for dependency `windows_x86_64_gnullvm`: 0.48.5, 0.52.6
multiple versions for dependency `windows_x86_64_msvc`: 0.48.5, 0.52.6
multiple versions for dependency `winnow`: 0.5.40, 0.6.13
multiple versions for dependency `base64`: 0.13.1, 0.21.7, 0.22.1
multiple versions for dependency `heck`: 0.3.3, 0.4.1, 0.5.0
multiple versions for dependency `regex-automata`: 0.1.10, 0.4.7
```

**Review by iamrecursion — COMMENTED** [review]
Looks good other than the comments.

**Review by iamrecursion — APPROVED** [review]
LGTM!

**Pull request merged**